### ISSUE TITLE
Group Volume Inconsistencies

### ISSUE THREAD

**Issue by davidjmeier** [avantrec/soco-cli#27]
First of all, thanks for such a great utility!  This could be on my end, but trying to confirm how to validate.

I'm using cached discovery of my speakers (6 in total) and am trying to implement some light automation to set the group, and normalize the volume within the group.

I'm using `Python 3.10.2` via a clean venv:
```
# pip list
Package            Version
------------------ ---------
anyio              3.5.0
appdirs            1.4.4
asgiref            3.5.0
certifi            2021.10.8
charset-normalizer 2.0.11
click              8.0.3
fastapi            0.73.0
h11                0.13.0
idna               3.3
ifaddr             0.1.7
pip                22.0.3
pydantic           1.9.0
rangehttpserver    1.2.0
requests           2.27.1
setuptools         60.7.1
sniffio            1.2.0
soco               0.26.0
soco-cli           0.4.36
starlette          0.17.1
tabulate           0.8.9
typing_extensions  4.0.1
urllib3            1.26.8
uvicorn            0.17.3
xmltodict          0.12.0
```

The problem I'm having is that if I leverage:
`# sonos -l 10.x.x.65 group-vol 25`

...I'd expect the group bound to all be set to a volume level of 25.

I'm starting out with this:
```
# sonos -l _all_ sysinfo 
Office-3: 
Report generated on: 2022-02-03 21:52 UTC (Thursday)

Zone Name    IP Address    Visible    CoOrd    CoOrd IP      Vol.   Mute    State    Model Name    Model No.    HW Version    SW Version
-----------  ------------  ---------  -------  -----------  ------  ------  -------  ------------  -----------  ------------  -----------------
Dining-5     10.x.x.64   Yes        No       10.x.x.65    7     Off     STOPPED  Play:5        S6           1.13.1.7-1.2  67.1-25031 (14.0)
Kitchen-N-1  10.x.x.63   Yes        No       10.x.x.65    4     Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Kitchen-S-1  10.x.x.60   Yes        No       10.x.x.65    4     Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Living-1     10.x.x.61   Yes        No       10.x.x.65    4     Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Master-1     10.x.x.62   Yes        No       10.x.x.65    4     Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Office-3     10.x.x.65   Yes        Yes                     7     Off     STOPPED  Play:3        S3           1.8.1.3-1.0   67.1-25031 (14.0)

Sonos model numbers present: S1, S3, S6.
Device counts: 6 total Sonos device(s), 3 unique model(s).
```

But when issue this:
`# sonos -l 10.x.x.65 group_vol 50`

And then reissue to validate volume levels I end up with something far more random like this:
```
# sonos -l _all_ sysinfo           
Office-3: 
Report generated on: 2022-02-03 21:55 UTC (Thursday)

Zone Name    IP Address    Visible    CoOrd    CoOrd IP      Vol.   Mute    State    Model Name    Model No.    HW Version    SW Version
-----------  ------------  ---------  -------  -----------  ------  ------  -------  ------------  -----------  ------------  -----------------
Dining-5     10.x.x.64   Yes        No       10.x.x.65    56    Off     STOPPED  Play:5        S6           1.13.1.7-1.2  67.1-25031 (14.0)
Kitchen-N-1  10.x.x.63   Yes        No       10.x.x.65    46    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Kitchen-S-1  10.x.x.60   Yes        No       10.x.x.65    46    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Living-1     10.x.x.61   Yes        No       10.x.x.65    46    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Master-1     10.x.x.62   Yes        No       10.x.x.65    46    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Office-3     10.x.x.65   Yes        Yes                     59    Off     STOPPED  Play:3        S3           1.8.1.3-1.0   67.1-25031 (14.0)

Sonos model numbers present: S1, S3, S6.
Device counts: 6 total Sonos device(s), 3 unique model(s).
```

If I go about this individually - the volumes fall in line as expected:
```
# sonos -l 10.x.x.60 vol 50
# sonos -l 10.x.x.61 vol 50
# sonos -l 10.x.x.62 vol 50
# sonos -l 10.x.x.63 vol 50
# sonos -l 10.x.x.64 vol 50
# sonos -l 10.x.x.65 vol 50
```

And now look at the system output again, I get what I'm after:
```
# sonos -l _all_ sysinfo     
Office-3: 
Report generated on: 2022-02-03 21:58 UTC (Thursday)

Zone Name    IP Address    Visible    CoOrd    CoOrd IP      Vol.   Mute    State    Model Name    Model No.    HW Version    SW Version
-----------  ------------  ---------  -------  -----------  ------  ------  -------  ------------  -----------  ------------  -----------------
Dining-5     10.x.x.64   Yes        No       10.x.x.65    50    Off     STOPPED  Play:5        S6           1.13.1.7-1.2  67.1-25031 (14.0)
Kitchen-N-1  10.x.x..63   Yes        No       10.x.x.65    50    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Kitchen-S-1  10.x.x.60   Yes        No       10.x.x.65    50    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Living-1     10.x.x.61   Yes        No       10.x.x.65    50    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Master-1     10.x.x.62   Yes        No       10.x.x.65    50    Off     STOPPED  Play:1        S1           1.8.3.7-1.0   67.1-25031 (14.0)
Office-3     10.x.x.65   Yes        Yes                     50    Off     STOPPED  Play:3        S3           1.8.1.3-1.0   67.1-25031 (14.0)

Sonos model numbers present: S1, S3, S6.
Device counts: 6 total Sonos device(s), 3 unique model(s).
```

Curious what I'm doing incorrectly, I haven't looked at the actual code yet - but wanted to see if this may be expected for some reason before I do.

Thanks, again, for the awesome utility!

**Comment by pwt** [comment]
Hi David, thanks for your kind comments, and for raising the issue.

The `group_volume` action is poorly described in my documentation. I'll correct this.

Behind the scenes, the action just uses the underlying Sonos primitives for adjusting the group volume. What this means is that if two or more speakers are grouped, if you adjust the group volume, then the specific adjustment for each individual speaker in the group will be 'scaled' depending on its current individual volume. This is the same behaviour as when using the volume slider in the Sonos app, when controlling grouped speakers: the 'group' slider will change the volumes of the individual speakers by different amounts.

Having said that, the values you're seeing don't look quite as I would expect, so I'll have to do a little experimentation. The algorithm Sonos uses for these group volume adjustments can seem a little ... odd.

Anyway, the bottom line for your use case, where you want all grouped speakers to be set to the same volume, you will need to set the speaker volumes individually, as your experiment confirms. However, it shouldn't be hard to implement a new action that works in the manner you were expecting, and I think it would be a good enhancement. 

I'll take a look at that tomorrow.

**Comment by pwt** [comment]
In SoCo-CLI v0.4.37 I've added a new `group_volume_equalise` action that will set all speakers in a group to the same specified volume, e.g.:

`sonos -l Dining-5 gve 50` will set the volume of all speakers in the same group as `Dining-5` to 50% volume.

The results you reported when using `group-vol` indicated that this action wasn't working as I would expect. This turns out to be a [bug](https://github.com/SoCo/SoCo/issues/887) in the underlying SoCo library (https://github.com/SoCo/SoCo/issues/887), which I've now [fixed](https://github.com/SoCo/SoCo/pull/888) and released. SoCo-CLI will pick up the patch release automatically when you update. Thanks for spotting this issue.

I've also clarified the documentation for the `group-volume` action.

Let me know if you have any questions or encounter any issues.
